rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadataFlyweight.java

```diff
@@ -0,0 +1,492 @@
+        final int metadataLength = compositeMetadata.readUnsignedMedium();
+        if (compositeMetadata.isReadable(metadataLength)) {
+          ByteBuf metadata =
+              retainSlices
+                  ? compositeMetadata.readRetainedSlice(metadataLength)
+                  : compositeMetadata.readSlice(metadataLength);
+          return new ByteBuf[] {mime, metadata};
+        } else {
+          return METADATA_MALFORMED;
+        }
+      } else {
+        return METADATA_MALFORMED;
+      }
+    }
+    return METADATA_BUFFERS_DONE;
+  }
+
+  /**
+   * Decode a {@code byte} compressed mime id from a {@link ByteBuf}, assuming said buffer properly
+   * contains such an id.
+   *
+   * <p>The buffer must have exactly one readable byte, which is assumed to have been tested for
+   * mime id encoding via the {@link #STREAM_METADATA_KNOWN_MASK} mask ({@code firstByte &
+   * STREAM_METADATA_KNOWN_MASK) == STREAM_METADATA_KNOWN_MASK}).
+   *
+   * <p>If there is no readable byte, the negative identifier of {@link
+   * WellKnownMimeType#UNPARSEABLE_MIME_TYPE} is returned.
+   *
+   * @param mimeBuffer the buffer that should next contain the compressed mime id byte
+   * @return the compressed mime id, between 0 and 127, or a negative id if the input is invalid
+   * @see #decodeMimeTypeFromMimeBuffer(ByteBuf)
+   */
+  public static byte decodeMimeIdFromMimeBuffer(ByteBuf mimeBuffer) {
+    if (mimeBuffer.readableBytes() != 1) {
+      return WellKnownMimeType.UNPARSEABLE_MIME_TYPE.getIdentifier();
+    }
+    return (byte) (mimeBuffer.readByte() & STREAM_METADATA_LENGTH_MASK);
+  }
+
+  /**
+   * Decode a {@link CharSequence} custome mime type from a {@link ByteBuf}, assuming said buffer
+   * properly contains such a mime type.
+   *
+   * <p>The buffer must at least have two readable bytes, which distinguishes it from the {@link
+   * #decodeMimeIdFromMimeBuffer(ByteBuf) compressed id} case. The first byte is a size and the
+   * remaining bytes must correspond to the {@link CharSequence}, encoded fully in US_ASCII. As a
+   * result, the first byte can simply be skipped, and the remaining of the buffer be decoded to the
+   * mime type.
+   *
+   * <p>If the mime header buffer is less than 2 bytes long, returns {@code null}.
+   *
+   * @param flyweightMimeBuffer the mime header {@link ByteBuf} that contains length + custom mime
+   *     type
+   * @return the decoded custom mime type, as a {@link CharSequence}, or null if the input is
+   *     invalid
+   * @see #decodeMimeIdFromMimeBuffer(ByteBuf)
+   */
+  @Nullable
+  public static CharSequence decodeMimeTypeFromMimeBuffer(ByteBuf flyweightMimeBuffer) {
+    if (flyweightMimeBuffer.readableBytes() < 2) {
+      return null;
+    }
+    // the encoded length is assumed to be kept at the start of the buffer
+    // but also assumed to be irrelevant because the rest of the slice length
+    // actually already matches _decoded_length
+    flyweightMimeBuffer.skipBytes(1);
```

Comment:
Does this adjust the readIndex? One the RSocket frames do is not adjust the read index to the call because it leads to weird problems when use the frames else.

Also I thought the first  bit not the first byte was what in was what determined length- so if the first bit is 0 do you need to skip anything?

Reply:
it indeed increases the `readerIndex()`. the first `bit` encodes the flag wether or not the remaining 7 bits represent a _length_ or a _mime id_ (if I interpreted the spec correctly, cc @nebhale). But since the previous step in decoding produced a buffer slice that has exactly `stringMimeTypeLength + 1` `readableBytes()`, we can assume this whole byte is A) starting with `1` and B) irrelevant to decode again.

The idea behind including this extra byte in the buffer was to avoid ambiguity between 2 possible 1-byte cases: 1 byte representing a compressed mime id vs 1 byte representing a 1-char mime string. By including it, the minimum length for the string case is 2 bytes, and thus we can differentiate by simply looking at `readableBytes()`.

Reply:
note that I assumed it was ok to move the `readerIndex()` when decoding and slicing?
(I rely on said index having moved when further decoding the next entry in the composite metadata buffer)

Reply:
Generally we've found its not okay to leave it changed once you're down with the flyweight. If you look at the RSocket frames they all reset the index afterwards to leave the bytebuf in the same state they found it.